packages/wallet/src/KeyManagement/LedgerKeyAgent.ts

```diff
@@ -0,0 +1,166 @@
+    activeTransport,
+    devicePath = ''
+  }: CreateTransportProps): Promise<TransportType> {
+    if (communicationType === CommunicationType.Node) {
+      return await TransportNodeHid.open(devicePath);
+    }
+    return await (activeTransport && activeTransport instanceof TransportWebHID
+      ? TransportWebHID.open(activeTransport.device)
+      : TransportWebHID.request());
+  }
+
+  static async createDeviceConnection(activeTransport: Transport): Promise<DeviceConnection> {
+    const deviceConnection = new DeviceConnection(activeTransport);
+    // Perform app check to see if device can respond
+    await deviceConnection.getVersion();
+    return deviceConnection;
+  }
+
+  static async establishDeviceConnection(
+    communicationType: CommunicationType,
+    devicePath?: string
+  ): Promise<DeviceConnection> {
+    let transport;
+    try {
+      transport = await LedgerKeyAgent.createTransport({ communicationType, devicePath });
+      if (!transport || !transport.deviceModel) {
+        throw new TransportError('Transport failed');
+      }
+      const isSupportedLedgerModel = transport.deviceModel.id === 'nanoS' || transport.deviceModel.id === 'nanoX';
+      if (!isSupportedLedgerModel) {
+        throw new TransportError('Ledger device model not supported');
+      }
+      return await LedgerKeyAgent.createDeviceConnection(transport);
+    } catch (error: any) {
+      if (error.message.includes('cannot open device with path')) {
+        throw new TransportError('Connection already established', error);
+      }
+      // If transport is established we need to close it so we can recover device from previous session
+      if (transport) {
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        transport.close();
+      }
+      throw error;
+    }
+  }
+
+  static async checkDeviceConnection(
+    communicationType: CommunicationType,
+    deviceConnection?: DeviceConnection
+  ): Promise<DeviceConnection> {
+    try {
+      if (!deviceConnection) {
+        return await LedgerKeyAgent.establishDeviceConnection(communicationType);
+      }
+      // Create / Check device connection with currently active transport
+      return await LedgerKeyAgent.createDeviceConnection(deviceConnection.transport);
+    } catch (error: any) {
+      // Device disconnected -> re-establish connection
+      if (error.name === 'DisconnectedDeviceDuringOperation') {
+        return await LedgerKeyAgent.establishDeviceConnection(communicationType);
+      }
+      throw error;
+    }
+  }
+
+  static async getXpub({
+    deviceConnection,
+    communicationType,
+    accountIndex
+  }: GetXpubProps): Promise<Cardano.Bip32PublicKey> {
+    const recoveredDeviceConnection = await LedgerKeyAgent.checkDeviceConnection(communicationType, deviceConnection);
+    const derivationPath = `1852'/1815'/${accountIndex}'`;
+    const extendedPublicKey = await recoveredDeviceConnection.getExtendedPublicKey({
```

Comment:
Should probably throw AuthenticationError if user refuses/cancels the operation on the device